proposal:kurator cd (#489)
kurator-dev/kurator: docs/proposals/kuratorcd/kurator-cd.md

---
title: Kurator's Continuous Delivery design 
authors:

> - "@LiZhenCheng9527" # Authors' GitHub accounts here.

Reviewer: ```suggestion
- @LiZhenCheng9527 # Authors' GitHub accounts here.
```

### Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users.
-->

With the increase in project size and complexity and the development of cloud computing technology, the CI/CD process has been proposed.

CI/CD has many advantages such as increased development efficiency, improved quality, more reliable deployment and better continuous learning and improvement, which is more suitable for today's software development process.

Kurator is an open source distributed cloud native suite that provides users with a one-stop open source solution for distributed cloud native scenarios.

Therefore, CI/CD as an important feature of cloud native usage scenarios, Kurator needs to provide relevant functional support to achieve the vision of Kurator unified configuration distribution.

#### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

Unified configuration distribution only requires the user to declare the desired API state in one place, and Kurator will automatically handle all subsequent operations.

> In Kurator, you can choose to distribute the application with the same configuration to a specific single or multiple clusters for verification.

Reviewer: make sure this has something todo with this proposal

> - **unified Continuous Delivery**

Reviewer: Captized.

> - **unified Continuous Delivery**

Reviewer: I can not understand Unified CD. IMO you should naming it to align with your title

Here's the preliminary design for the Unified Continuous Delivery:

```console
// ApplicationSyncPolicy defines the configuration to sync an artifact.
// Only oneof `kustomization` or `helm` can be specified to manage application sync.
// ApplicationSyncPolicy distributes the Continuous Delivery configuration 
// at the same time as the application deployment, if needed. 
type ApplicationSyncPolicy struct {
    // Name defines the name of the sync policy.
    // If unspecified, a name of format `<application name>-<index>` will be generated.
    // +optional
    Name string `json:"name,omitempty"`

    // Kustomization defines the configuration to calculate the desired state
    // from a source using kustomize.
    // +optional
    Kustomization *Kustomization `json:"kustomization,omitempty"`
    // HelmRelease defines the desired state of a Helm release.
    // +optional
    Helm *HelmRelease `json:"helm,omitempty"`

    // Destination defines the destination for the artifact.
    // If specified, it will override the destination specified at Application level.
    // +optional

>     Destination *ApplicationDestination `json:"destination"`

Reviewer: hide these fields

    // Rollout defines the rollout Configurations to be used.
    // If specified, a uniform Continuous Delivery policy is configured for this installed object.
    // +optional
    Rollout *RolloutConfig `json:"rolloutPolicy,omitempty"`
}

type RolloutConfig struct {
    // Testload defines Whether to install testload for users. Default is true.
    // testload generates traffic during canary analysis.
    // If set it to false, user need to install the testload himself.
    // If set it to true or leave it blank, Kurator will install the flagger's testload.
    // +optional
    TestLoad bool `json:"testLoad,omitempty"`

    // Kurator only supports istio for now.
    // New Provider will be added later.
    // +optional
    TrafficRoutingProvider string `json:"trafficRoutingProvider,omitempty"`

    // Workload specifies what workload to deploy the test to. 
    // Workload of type deployment or daemonSet.
    Workload *WorkloadReference `json:"workload"`

    // Port of the Workload which want to test.

>     Port int32 `json:"port"`

Reviewer: confused about this.

So if there are two ports defined in the service, user can specify one here?

>     // ServiceName holds the name of a service which selects pods with Workload.

Reviewer: ```suggestion
    // ServiceName holds the name of a service which matches the `workload`.
```

    // canary deployment to make progress before it is considered to be failed.
    // Defaults to 600s.
    // +optional
    ProgressDeadlineSeconds *int32 `json:"progressDeadlineSeconds,omitempty"`

    // SkipAnalysis promotes the canary without analyzing it
    // +optional
    SkipTrafficAnalysis bool `json:"skipTrafficAnalysis,omitempty"`

    // Restore resources to initial state when deleting canary resources.
    // Use of the revertOnDeletion property should be enabled
    // when you no longer plan to rely on Kurator for deployment management.
    // +optional
    RevertOnDeletion bool `json:"revertOnDeletion,omitempty"`

    // Suspend, if set to true will suspend the Canary, disabling any canary runs
    // regardless of any changes to its target, services, etc. Note that if the
    // Canary is suspended during an analysis, its paused until the Canary is uninterrupted.
    // +optional
    Suspend bool `json:"suspend,omitempty"`
}
```

TargetObjectReference contains enough information to let you locate the typed referenced object in the same namespace. The two types of Kind now supported are `Deployment` and `DaemonSet`.

```console

> type TargetWorkloadReference struct {

Reviewer: donot we need a group field?

Check we have may objectref

    // API version of the referent
    // +optional
    APIVersion string `json:"apiVersion,omitempty"`

    // Kind of the referent
    // +optional
    Kind string `json:"kind,omitempty"`

    // Name of the referent
    Name string `json:"name"`
}
```

Kurator will create a VirtualService resource based on the configuration in `VirtualServiceConfig` to distribute traffic.

```console
// Note: refer to https://github.com/fluxcd/flagger/blob/main/pkg/apis/flagger/v1beta1/canary.go
type TrafficRoutingConfig struct {
    // Timeout of the HTTP or gRPC request.
    // Timeout in upstream response time.
    // +optional
    Timeout string `json:"timeout,omitempty"`

    // Gateways attached to the generated Istio virtual service.
    // Defaults to the internal mesh gateway.
    // +optional
    Gateways []string `json:"gateways,omitempty"`

    // Iterations defines the number of checks to run for A/B Testing and Blue/Green
    // Note: Flagger determines whether blue-green or A/B related processing is required based on 
    // the presence or absence of content in the Iterations field. 
    // So can't configure Iterations and CanaryStrategy at the same time.
    // +optional

>     Iterations int `json:"iterations,omitempty"`

Reviewer: This is vague, please rename

---
title: Kurator's Continuous Delivery design 
authors:
- "@LiZhenCheng9527" # Authors' GitHub accounts here.
reviewers:
approvers:

creation-date: 2023-11-20

---

## Kurator's Continuous Delivery design

<!--
This is the title of your KEP. Keep it short, simple, and descriptive. A good
title can help communicate what the KEP is and should be considered as part of
any review.
-->

### Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. 

A good summary is probably at least a paragraph in length.
-->

> Kurator, as an open-source distributed cloud-native platform, has been pivotal in aiding users to construct their distributed cloud-native infrastructure, thereby facilitating enterprise digital transformation.

Reviewer: Do not repeat this, too general.

To further enhance its functionality, this proposal designs Kurator's Continuous Delivery feature to meet users' needs for automated distribution of verified code.

By integrating Flagger, we aim to provide users with reliable, fast and unified Continuous Delivery, enabling them to easily distribute code across multiple clusters.

> Base on Flagger, Kurator also offers A/B, Blue/Green and Canary distribution options. Meet the needs of the test verification release.

Reviewer: `test verification release` is this a commonly used phrase?

### Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users.
-->

With the increase in project size and complexity and the development of cloud computing technology, the CI/CD process has been proposed.

CI/CD has many advantages such as increased development efficiency, improved quality, more reliable deployment and better continuous learning and improvement, which is more suitable for today's software development process.

Kurator is an open source distributed cloud native suite that provides users with a one-stop open source solution for distributed cloud native scenarios.

Therefore, CI/CD as an important feature of cloud native usage scenarios, Kurator needs to provide relevant functional support to achieve the vision of Kurator unified configuration distribution.

#### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> Unified configuration distribution only requires the user to declare the desired API state in one place, and Kurator will automatically handle all subsequent operations.

Reviewer: why is the goal for `Unified configuration distribution`

> Kurator deploys different daemons on different nodes in the cluster with different rules for configuration distribution based on the declarations in the Spec.

Reviewer: ? can not get  the intention here

In Kurator, you can choose to distribute the application with the same configuration to a specific single or multiple clusters for verification.

- **unified Continuous Delivery**
    - Supports unified configuration of releases for multiple clusters. Achieve the deployment configuration of the application to be distributed to the specified single or multiple clusters.

>     - Support A/B, Blue/Green, Canary three release validation rules configuration.

Reviewer: These three are all deploy strategies, donot mention validation? It is confusing to users

    - Supports automatic rollback when release validation fails.

#### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

> - **Traffic distribution tools that Flagger can use other than istio are not supported** While Flagger is able to support a wide range of traffic distribution tools including istio, nginx for grey scale releases. However, Kuraotr currently only supports unified installation of istio in multiple clusters, and Kurator may implement unified installation of other traffic distribution tools in the future.

Reviewer: ```suggestion
- **Traffic distribution tools other than istio are not supported** While Flagger is able to support a wide range of traffic distribution tools including istio, nginx for grey scale releases. However, Kurator currently only supports unified management of istio across clusters. Kurator may implement other traffic distribution tools in the future.
```

### Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->
The purpose of this proposal is to introduce a unified Continuous Delivery for Kurator that supports A/B, Blue/Green, and Canary.The main objectives of this proposal are as follows:

> Custom Resource Definitions (CRDs): Design CRDs to enable Uniform Continuous Delivery These CRDs will provide a structured approach to defining clusters and different configuration distribution rules to enable uniform configuration distribution.

Reviewer: API Design

Reviewer: CRD is just a present type

> Fleet-Manager Implementation: The Cluster Manager component will be responsible for monitoring the CRDs and performing the defined functions. It will install Flagger on the clusters and handle potential errors or anomalies to ensure smooth operation.

Reviewer: Not fleet-manager implementation. You are implementing CD, IIRC, you want to introduce a new component. So it is the implement of XXXX

Here's the preliminary design for the Unified Continuous Delivery:

```console
// ApplicationSyncPolicy defines the configuration to sync an artifact.
// Only oneof `kustomization` or `helm` can be specified to manage application sync.
// ApplicationSyncPolicy distributes the Continuous Delivery configuration 
// at the same time as the application deployment, if needed. 
type ApplicationSyncPolicy struct {
    // Name defines the name of the sync policy.
    // If unspecified, a name of format `<application name>-<index>` will be generated.
    // +optional
    Name string `json:"name,omitempty"`

    // Kustomization defines the configuration to calculate the desired state
    // from a source using kustomize.
    // +optional
    Kustomization *Kustomization `json:"kustomization,omitempty"`
    // HelmRelease defines the desired state of a Helm release.
    // +optional
    Helm *HelmRelease `json:"helm,omitempty"`

    // Destination defines the destination for the artifact.
    // If specified, it will override the destination specified at Application level.
    // +optional

>     Destination *ApplicationDestination `json:"destination"`

Reviewer 2: BTW, `Destination` may be refactored in the future. See https://github.com/kurator-dev/kurator/issues/439.

    // Delivery defines the Continuous Delivery Configurations to be used.
    // If specified, a uniform Continuous Delivery policy is configured for this installed object.
    // +optional

>     Delivery *DeliveryConfig `json:"deliveryPolicy"`

Reviewer: Call `Rollout`?

}

> type DeliveryConfig struct {

Reviewer: RollOutConfig

    // Autoscale defines minreplicas and maxreplicas of horizontal Pod Autoscaler.
    // +optional

>     Autoscale *AutoscaleRef `json:"autoScale,omitempty"`

Reviewer: It is not clear, how this will influence the rolling out

>     // Whether or not the testload is installed by the user. The default is false.

Reviewer: ```suggestion
    // Whether or not the testloader is installed by the user. The default is false.
```

>     // If set it to ture, user need to install the testload himself.

Reviewer: ```suggestion
    // If set to true, user needs to install the testloader himself.
```

Reviewer: I think the semantics is negative. 

We can make it: testLoader will be installed when this is true

    // If set it to false or leave it blank, Kurator will install the flagger's testload.
    // +optional

>     TestLoad bool `json:"testLoad"`

Reviewer 2: nit, omitempty

>     // DeliveryPolicy defines the Release Strategy of Object

Reviewer: ``
```suggestion
    // DeliveryPolicy defines the release strategy of XXXX
```

Reviewer: It is not object, it is deployment

    DeliveryPolicy *DeliveryPolicy `json:"deliveryPolicy"`
}
```

> Kurator generates horizontalpodautoscaler.autoscaling resources based on `autoscale` configurations as stable and to-be-verified versions. For instance tuning during testing

Author: ```suggestion
Kurator generates horizontalpodautoscaler.autoscaling resources based on `autoscale` configurations as stable and to-be-verified versions. Make adjustments during testing
```

```concle
type AutoscaleRef struct {
    // +optional
    MinReplicas *int32 `json:"minReplicas,omitempty"`

    // +optional
    MaxReplicas *int32 `json:"maxReplicas,omitempty"`
}
```

`Testload` indicates whether the user wants to install the test traffic load themselves. If you don't want to install the testload yourself, Kurator will install flagger's testload by default.

`DeliveryPolicy` defines the Continuous Delivery configuration for this installation object. Although in Kurator is not detailed into the canary release, A/B test and blue-green test. However, users can freely configure their own traffic validation policies through the DeliveryPolicy. Complete the release test.

```console
// Note: refer to https://github.com/fluxcd/flagger/blob/main/pkg/apis/flagger/v1beta1/canary.go
type DeliveryPolicy struct {
    // Kurator only supports istio for now.
    // New Provider will be added later.
    // +optional
    TrafficRoutingProvider string `json:"trafficRoutingProvider,omitempty"`

    // TargetObject specifies what object to deploy the test to. 
    // Objects of type deployment or daemonSet.

>     TargetObject *TargetObjectReference `json:"targetObject"`

Reviewer: Move to the upper struct DeliveryConfig?

This must be specified

Reviewer: s/TargetObject/WorkloadRef ?

    // +optional
    SkipTrafficAnalysis bool `json:"skipTrafficAnalysis,omitempty"`

    // Restore resources to initial state when deleting canary resources.
    // Use of the revertOnDeletion property should be enabled
    // when you no longer plan to rely on Kurator for deployment management.
    // +optional
    RevertOnDeletion bool `json:"revertOnDeletion,omitempty"`

    // Suspend, if set to true will suspend the Canary, disabling any canary runs
    // regardless of any changes to its target, services, etc. Note that if the
    // Canary is suspended during an analysis, its paused until the Canary is unsuspended.
    // +optional
    Suspend bool `json:"suspend,omitempty"`
}
```

TargetObjectReference contains enough information to let you locate the typed referenced object in the same namespace. The two types of Kind now supported are `Deployment` and `DaemonSet`.

```concole
// Note: refer to https://github.com/fluxcd/flagger/blob/main/pkg/apis/flagger/v1beta1/canary.go
type TargetObjectReference struct {
    // API version of the referent
    // +optional
    APIVersion string `json:"apiVersion,omitempty"`

    // Kind of the referent
    // +optional

>     Kind string `json:"kind,omitempty"`

Reviewer 2: If 'kind' is optional, how can we specify or identify an object

    // Name of the referent
    Name string `json:"name"`
}
```

The ServiceConfig is the configuration of the service that is generated according to the object specified in the TargetObject. The main thing is to specify the `Port` and `targetPort`.

```console
// Note: refer to https://github.com/fluxcd/flagger/blob/main/pkg/apis/flagger/v1beta1/canary.go

> type ServiceConfig struct {

Reviewer: is this used for canary svc generation? If so why does not it use the original svc's spec.

>     // Name of the virtual Kubernetes generated by Flagger.

Reviewer: `virtual Kubernetes` means what?

    // Defaults to DeliveryPolicy.TargetObject.Name
    // +optional
    Name string `json:"name,omitempty"`

    // Port of the generated Kubernetes service.
    Port int32 `json:"port"`

    // Port name of the generated Kubernetes service.

>     // Defaults to http

Reviewer 2: what does "Defaults to http" means, the defaults PortName is "http"?

Author: Yes. More information you can refer to [canary-service](https://docs.flagger.app/v/main/usage/how-it-works#canary-service)

    // Defaults to CanaryService.Port
    // +optional
    TargetPort intstr.IntOrString `json:"targetPort,omitempty"`

    // AppProtocol of the service.
    // https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol
    // +optional
    AppProtocol string `json:"appProtocol,omitempty"`

    // PortDiscovery adds all container ports to the generated Kubernetes service.
    // Defaults to true
    PortDiscovery bool `json:"portDiscovery"`
}
```

Kurator will create a VirtualService resource based on the configuration in `VirtualServiceConfig` to distribute traffic.

```console
// Note: refer to https://github.com/fluxcd/flagger/blob/main/pkg/apis/flagger/v1beta1/canary.go
type VirtualServiceConfig struct {
    // Timeout of the HTTP or gRPC request.
    // +optional
    Timeout string `json:"timeout,omitempty"`

    // Gateways attached to the generated Istio virtual service.
    // Defaults to the internal mesh gateway.
    // +optional
    Gateways []string `json:"gateways,omitempty"`

    // Hosts attached to the generated Istio virtual service.
    // Defaults to the service name
    // +optional
    Hosts []string `json:"hosts,omitempty"`

>     // If enabled, Flagger would generate Istio VirtualServices without hosts and gateway,

Reviewer 2: nit, Annotations should start with the parameter name, followed by its description. For example, 'Delegation [description of Delegation]'

    // making the service compatible with Istio delegation.
    // If delegation is enabled, Flagger makes the service compatible with Istio delegation,
    // enabling configuration of traffic within the cluster
    // Note that pilot env `PILOT_ENABLE_VIRTUAL_SERVICE_DELEGATE` must also be set.
    // +optional
    Delegation bool `json:"delegation,omitempty"`

    // TrafficPolicy attached to the generated Istio destination rules.
    // +optional
    TrafficPolicy *istiov1alpha3.TrafficPolicy `json:"trafficPolicy,omitempty"`

>     // URI match conditions for the generated service.

Reviewer 2: ```suggestion
   // Match specifies URI match conditions for the generated service.
```

    // +optional
    Match []istiov1alpha3.HTTPMatchRequest `json:"match,omitempty"`

    // Retries policy for the generated virtual service.
    // +optional
    Retries *istiov1alpha3.HTTPRetry `json:"retries,omitempty"`

    // Headers operations for the generated Istio virtual service.
    // e.g.
    // headers:
    //   request:
    //     add:
    //       x-some-header: "value"
    // +optional
    Headers *istiov1alpha3.Headers `json:"headers,omitempty"`

    // Cross-Origin Resource Sharing policy for the generated Istio virtual service.
    // e.g.
    // corsPolicy:
    //   allowHeaders:
    //   - x-some-header
    //   allowMethods:
    //   - GET
    //   allowOrigin:
    //   - example.com
    //   maxAge: 24h
    // +optional
    CorsPolicy *istiov1alpha3.CorsPolicy `json:"corsPolicy,omitempty"`

    // Primary is the metadata to add to the primary service.
    // +optional
    Primary *CustomMetadata `json:"primary,omitempty"`

    // Canary is the metadata to add to the canary service.
    // +optional

>     Canary *CustomMetadata `json:"canary,omitempty"`

Reviewer: Not sure i understand. Is this annotation?
